slidechain/export.go

```diff
@@ -271,117 +368,64 @@ func BuildExportTx(ctx context.Context, asset xdr.Asset, amount, inputAmt int64,
+	b.Op(op.Finalize)                                                       // con stack: sigchecker
+	prog1 := b.Build()
+	vm, err := txvm.Validate(prog1, 3, math.MaxInt64, txvm.StopAfterFinalize)
 	if err != nil {
-		return nil, errors.Wrap(err, "building tx")
-	}
-	if inputAmt > amount {
-		txresult := txresult.New(tx)
-		output := txresult.Outputs[0].Value
-		log.Printf("output: assetid %x amount %x anchor %x", output.AssetID.Bytes(), output.Amount, output.Anchor)
-	}
-	return tx, nil
-}
+		return nil, errors.Wrap(err, "computing transaction ID")
+	}
+	sigProg := standard.VerifyTxID(vm.TxID)
+	msg := append(sigProg, anchor...)
+	sig := ed25519.Sign(prv, msg)
+	b.PushdataBytes(sig).Op(op.Put)
+	b.PushdataBytes(sigProg).Op(op.Put)
+	b.Op(op.Call)
 
-// IsExportTx returns whether or not a txvm transaction matches the slidechain export tx format.
-//
-// Expected log is:
-// For an export that fully consumes the input:
-// {"I", ...}
-// {"L", ...}
-// {"X", vm seed, inputAmount, asset id, anchor}
-// {"L", vm seed, refdata}
-// {"R", ...} timerange
-// {"L", ...}
-// {"F", ...}
-//
-// For an export that partially consumes the input:
-// {"I", ...}
-// {"L", ...}
-// {"X", vm seed, inputAmount, asset id, anchor}
-// {"L", vm seed, refdata}
-// {"L", ...}
-// {"L", ...}
-// {"O", caller, outputid}
-// {"R", ...}
-// {"L", ...}
-// {"F", ...}
-func IsExportTx(tx *bc.Tx, asset xdr.Asset, inputAmt int64, temp, exporter string, seqnum int64) bool {
-	// The export transaction when we export the full input amount has seven operations, and when we export
-	// part of the input and output the rest back to the exporter, it has ten operations
-	if len(tx.Log) != 7 && len(tx.Log) != 10 {
-		return false
-	}
-	if tx.Log[0][0].(txvm.Bytes)[0] != txvm.InputCode {
-		return false
-	}
-	if tx.Log[1][0].(txvm.Bytes)[0] != txvm.LogCode {
-		return false
-	}
-	if tx.Log[2][0].(txvm.Bytes)[0] != txvm.RetireCode {
-		return false
-	}
-	if int64(tx.Log[2][2].(txvm.Int)) != inputAmt {
-		return false
-	}
-	assetBytes, err := asset.MarshalBinary()
+	prog2 := b.Build()
+	tx, err := bc.NewTx(prog2, 3, math.MaxInt64)
 	if err != nil {
-		return false
+		return nil, errors.Wrap(err, "making export tx")
 	}
-	assetXDR, err := xdr.MarshalBase64(asset)
-	if err != nil {
-		return false
-	}
-	wantAssetID := txvm.AssetID(importIssuanceSeed[:], assetBytes)
-	if !bytes.Equal(wantAssetID[:], tx.Log[2][3].(txvm.Bytes)) {
-		return false
-	}
-	if tx.Log[3][0].(txvm.Bytes)[0] != txvm.LogCode {
-		return false
-	}
-	ref := struct {
-		AssetXDR string `json:"asset"`
-		Temp     string `json:"temp"`
-		Seqnum   int64  `json:"seqnum"`
-		Exporter string `json:"exporter"`
-	}{
-		assetXDR,
-		temp,
-		seqnum,
-		exporter,
-	}
-	refdata, err := json.Marshal(ref)
-	if !bytes.Equal(refdata, tx.Log[3][2].(txvm.Bytes)) {
-		return false
-	}
-	// Beyond this, the two transactions diverge but must either finalize
-	// or output the remaining unconsumed input
-	return true
+	return tx, nil
```

Comment:
This and the preceding three lines can be collapsed to `return tx, errors.Wrap(err, ...)`.